### DISCUSSION TITLE
Yamaha NP-S2000 not working!

### DISCUSSION THREAD

**Discussion by Ice64zzz** [coffeegreg/YTuner#36] [Q&A] [answered]
Hello! Thank you for yoor work, but this program doesn't work with my Yamaha np-s2000 network player. I see list of radio station, can choose one of them, but after that progress bar on yamaha's display stunning and nothing happen. After that i can make return, change other station, bit no effect btw. I foungd only one version of YCast, which works with my NP-S2000 (https://github.com/jonnieZG/YCast). He told in discussion, that there is a problem with main line of YCast (Station UUIDs used by radiobrowser are translated into shorter strings, to work with some older radios)(https://github.com/milaq/YCast/pull/79/commits/c8ccc9649f9804050129ec277c8db7e3c2527b8d) Can you add this fix for older radios (such as options in cfg file for example) or send me compilled version for my yamaha NP-S2000. I'm not so good at programming, so i cannot do it by myself. This would help me a lot.

**Comment by TheBossME** [comment]
Hi, 

I knew about the issue you described with ycast.

How about using the default stations.ini or yaml file which is coming with installation of ytuner ?

Is that playing properly?

DNS Adress properly changed on your Yamaha device to the DNS Adress of the machine running ytuner ?

Disconnect Yamaha device from power plug and disconnect network cable. Device is maybe also caching information.

DNS & Gateway set properly on Yamaha device ?

Ytuner.ini and AVR.ini are the most important setting files.
Use debug=4 option and post the log result please.


Regards
Beatrice

**Reply by Ice64zzz** [reply]
Nope, I can see stations from stations.ini But when I choose one of them I see empty load bar on my display and it’s not playable. Network configuration (DNS and Gateway is good. This program and Milaq’s ycast not working with correct network configuration, but jonnieZG’s ycast works good on the same server. The problem is limitation of UUID variable in my np-s2000 (I guess). I can turn on debug=4 a bit later.

**Comment by coffeegreg** [comment]
Hi @Ice64zzz 
This YCast commit https://github.com/milaq/YCast/commit/c8ccc9649f9804050129ec277c8db7e3c2527b8d description is about:
> Station UUIDs used by radiobrowser are translated into shorter strings

I suspect that the problem with the NP-S2000 is not related to the operation of `Radiobrowser` itself and the UUIDs used only, but to the need for the AVR to support shorter strings as identifiers, because from your description it follows that you cannot correctly play stations from `stations.ini/yaml` files either. The identifier that YTuner generates is a composition of several elements that YCast does not have (e.g. bookmarks support), so in order not to destroy the current functionality of YTuner, the "shortening" of this identifier must look slightly different than in this YCast fork. 
I didn't even realize that some AVRs might have a problem with its size. :thinking: Anyway, good to know. :+1: 

First of all, I would like to see info from the YTuner log (debug=4). First, when you try to play radio station from `stations.ini/yaml` and second, when you try to play radio station from `Radiobrowser`. Then I will see what needs to be changed to make your Yamaha play with YTuner as well. :speaker: :notes: I imagine the whole thing as a new parameter in which the user sets the maximum size for this identifier. :thinking: 

I am currently working on version `1.2.4 beta` which will include some last fixes and adding a translator to YTuner ... yeah ... Beatrice inspired me :wink: :rocket: If you provide the logs I wrote about, I may also add a fix for NP-S2000. :thinking:

**Reply by Ice64zzz** [reply]
YTuner v1.2.3 Copyright (c) 2024 Greg P. (https://github.com/coffeegreg)
04.11.2024 20:30:21 : Inf : Starting services...
04.11.2024 20:30:21 : Inf : Successfully loaded 10 my stations.
04.11.2024 20:30:21 : Dbg : Database library : 3.45.0.
04.11.2024 20:30:21 : Inf : Checking local database.
04.11.2024 20:30:22 : Inf : Local database is ready.
04.11.2024 20:30:22 : Inf : DNS Service: listening on: 192.168.101.113:53.
04.11.2024 20:30:22 : Inf : Web Service: listening on: 192.168.101.113:80.
04.11.2024 20:30:31 : Dbg : DNS Query intercept : radioyamaha.vtuner.com.
04.11.2024 20:30:31 : Dbg : GET /setupapp/Yamaha/asp/BrowseXML/loginXML.asp?mac=9fad5cfe1dbca2ff078661f0253a34c1&dlang=eng&fver=W&start=1&howmany=8.
04.11.2024 20:30:37 : Dbg : DNS Query intercept : radioyamaha.vtuner.com.
04.11.2024 20:30:37 : Dbg : GET /ytuner/my_stations?ytuner=true&mac=9fad5cfe1dbca2ff078661f0253a34c1&dlang=eng&fver=W&start=1&howmany=8.
04.11.2024 20:30:42 : Dbg : DNS Query intercept : radioyamaha.vtuner.com.
04.11.2024 20:30:42 : Dbg : GET /ytuner/my_stations/Dance?ytuner=true&mac=9fad5cfe1dbca2ff078661f0253a34c1&dlang=eng&fver=W&start=1&howmany=8.
04.11.2024 20:30:45 : Dbg : DNS Query intercept : radioyamaha.vtuner.com.
04.11.2024 20:30:45 : Dbg : GET /setupapp/Yamaha/asp/BrowseXML/statxml.asp?mac=9fad5cfe1dbca2ff078661f0253a34c1&fver=W&id=MS_534A591AB545.
04.11.2024 20:30:46 : Dbg : GET /ytuner/icon?id=MS_534A591AB545.
04.11.2024 20:31:02 : Dbg : DNS Query intercept : radioyamaha.vtuner.com.
04.11.2024 20:31:02 : Dbg : GET /ytuner/my_stations?ytuner=true&mac=9fad5cfe1dbca2ff078661f0253a34c1&dlang=eng&fver=W&start=1&howmany=8.
04.11.2024 20:31:09 : Dbg : DNS Query intercept : radioyamaha.vtuner.com.
04.11.2024 20:31:09 : Dbg : GET /setupapp/Yamaha/asp/BrowseXML/loginXML.asp?mac=9fad5cfe1dbca2ff078661f0253a34c1&dlang=eng&fver=W&start=1&howmany=8.
04.11.2024 20:31:29 : Dbg : DNS Query intercept : radioyamaha.vtuner.com.
04.11.2024 20:31:29 : Dbg : GET /ytuner/radiobrowser?ytuner=true&mac=9fad5cfe1dbca2ff078661f0253a34c1&dlang=eng&fver=W&start=1&howmany=8.
04.11.2024 20:31:48 : Dbg : DNS Query intercept : radioyamaha.vtuner.com.
04.11.2024 20:31:48 : Dbg : GET /ytuner/radiobrowser/popular?ytuner=true&mac=9fad5cfe1dbca2ff078661f0253a34c1&dlang=eng&fver=W&start=1&howmany=8.
04.11.2024 20:31:53 : Dbg : DNS Query intercept : radioyamaha.vtuner.com.
04.11.2024 20:31:53 : Dbg : GET /setupapp/Yamaha/asp/BrowseXML/statxml.asp?mac=9fad5cfe1dbca2ff078661f0253a34c1&fver=W&id=RB_962CC6DF0601.
04.11.2024 20:31:54 : Dbg : GET /ytuner/icon?id=RB_962CC6DF0601.

**Comment by coffeegreg** [comment]
This is weird... :thinking: 
It looks like your NP-S2000 requests are correct and the radio station IDs are correct (15 characters) and there are no effects of "truncating" the ID string to shorter.
`MS_534A591AB545` is a valid ID of radio station from `My Stations` menu and `stations.ini/yaml` file. In this case there are no UUIDs you wrote about. In fact, YTuner from the very beginning does not use the full length of the UUID and the length of the station ID is 15 characters.
The strange thing is that your AVR keeps asking DNS for the radioyamaha.vtuner.com domain. It should only do this once at the beginning... :thinking:
It looks like your AVR can't connect to the selected radio station. It looks like you have DNS configured correctly on your AVR because your AVR sends requests to the YTuner machine and the YTuner DNS service listening on UDP 53 port catches all requests that lead to *ytuner.com domains ... but ... this multiple `DNS Query intercept` means that the AVR did not receive a valid IP address for `radioyamaha.vtuner.com` domain or does not cache it. :thinking: Please try to change `DNSServers` in `[DNSServer]` section of `ytuner.ini` file to the appropriate one(s) for your ISP. YTuner DNS service intercepts all `*vtuner.com` requests and for all the rest query the DNS server(s) which is/are entered in `DNSServers` setting of `ytuner.ini` file.
Another thing .. Are you absolutely sure you have checked YTuner, YCast and this YCast fork (https://github.com/jonnieZG/YCast) on the same machine, with the same config and the same `stations.yml` file?

**Reply by Ice64zzz** [reply]
GET /ytuner/icon?id=RB_962CC6DF0601. Works correctly so I can see logo of the station in my AVR’s web-client, but music cannot pass through. Is station logos approve that DNS works well? I tested completely the same configuration of YCast and YCast jonnieZG fork with manual DNS entry (ip host radioyamaha.vtuner.com [ycast_ip]) on my router. So YCast - doesn’t work, YCast jonnieZG fork - works well. For YTuner I’m using static parameters (ip, dns and gw) so it’s slightly different configuration. I didn’t try my own station (stations.ini) at YCast and at YCast jonnieZG fork, I tried only radiobrowser’s list. I’m always using 8.8.8.8 as DNS for my ISP connection, ignoring their DNS. It works without any problem. So, what should I do now?
Upd: my AVR asking DNS for the radioyamaha.vtuner.com domain when I click any button in AVR’s web menu (click to my stations or radiobrowser, click on station to play it, click back to menu) one click - one line “ DNS Query intercept : radioyamaha.vtuner.com.”
Upd2: will this topic help? -> https://github.com/milaq/YCast/pull/79 (https://github.com/milaq/YCast/pull/79#issuecomment-913634191 helps in my case too)

**Reply by coffeegreg** [reply, marked as answer]
> Is station logos approve that DNS works well?

No. YTuner downloads radio station logo images, converts and scales them to the correct format/size and stores them in its cache. So the station logo is taken directly from YTuner, not from the radio station's web service. Please read the entire contents of the `ytuner.ini` and `avr.ini` configuration files as @TheBossME advised. There are a lot of settings there that YCast doesn't have and that can be useful for you. Moreover, without proper configuration of some settings from these files, YTuner may not work properly in some conditions.

Let's do a test ... prepare a `stations.ini` file with the following content :
```
[Dance]
FFH - Eurodance=http://145.239.0.3/ffhchannels/hqeurodance.mp3
Splash 90s Dance=http://151.80.6.109/radio/8090/radio.mp3
```
As you can see I've replaced domain names with IP addresses. Try it and this test should give us the answer about DNS related problems.

>I’m always using 8.8.8.8 as DNS for my ISP connection, ignoring their DNS.

Please try with other DNS Servers. Do some tests .. change to 9.9.9.9 only with `DNSServers` setting in `ytuner.ini` file or try with default DNS Serwer of your ISP. The tests don't cost anything but can give us many answers. :+1: 

>Upd: my AVR asking DNS for the radioyamaha.vtuner.com domain when I click any button in AVR’s web menu (click to my stations or radiobrowser, click on station to play it, click back to menu) one click - one line “ DNS Query intercept : radioyamaha.vtuner.com.”

As I mentioned earlier, AVR tries to resolve the domain name `radioyamaha.vtuner.com` to an IP address but for some reason it doesn't get it and keeps trying every time.

>Upd2: will this topic help?

No. I don't think so.

**Reply by TheBossME** [reply]
These are my dns redirection settings on pihole.
I had also that issue with ycast. And it worked with Jonny's workaround.

But for me no way back to ycast..

address=/radioyamaha.vtuner.com/192.168.x.1
address=/radioyamaha2.vtuner.com/192.168.x.2
address=/primary1b.vtuner.com/192.168.x.1
address=/backup2.vtuner.com/192.168.x.2

**Reply by coffeegreg** [reply]
YTuner never used full length UUIDs as station IDs so this problem is not specific to YTuner. Besides, here we have a problem with NP-S2000 regardless of whether it tries to use `Radiobrowser` or `My Stations` which do not use radiobrowser specific UUIDs at all.

**Reply by TheBossME** [reply]
My guess is very much in the direction of a DNS setup problem with ytuner.

Have you disconnect/reconnect Yamaha device from power and network. 

I saw similar issue on my device after switching from ycast to ytuner.

**Comment by Ice64zzz** [comment]
Finally got it working! I don't know the exact reason, but I just installed YTuner on an external VDS server, disabled DNS in ytuner.ini, created DNS entry radioyamana.vtuner.com to external VDS server on my router and everything started working. I have a few small issues, such as lack of icons in np-s2000's web interface in mobile browsers (Chrome for desktop shows them) and the native yamaha mobile iOS app called NP Controller doesn't show the list of internet radio station (loading..) so i cannot use it. But this problems exists in YCast too.
Thank you for your advices and patience!  Confirmed working with Yamaha NP-S2000!
P.S.: Do you plan to do on-the-fly transcoding, for example from AAC to MP3? Another one of my wishes: the ability to see the radio quality before or after the title, for example "[128] Dance Wave!" on my yamaha's display.

**Reply by TheBossME** [reply]
I have filtered out aac stations and for mp3 files I filtered out all below 128. but, database is not properly updated.

The only proper way is analyzing the stream data, but this is not implemented with ytuner and also not at vtuner or ycast.

Also transcoding from aac to mp3 isn't actually a good idea. It needs a powerful real-time transcoding system. Greg and I discussed about that. When a mp3 stream is available for a station then I use the AAC filtering and didn't see that AAC station. FLAC to MP3 is also not a good idea ! 

We should think about additional hardware or other ways to play FLAC or AAC files. The DNLA functionality could be part of.

ytuner integration with DLNA features could maybe a way to do.

**Reply by coffeegreg** [reply]
> Finally got it working! 

I'm glad to know that YTuner also works properly for you. 😃 🔈 🎶

>  I don't know the exact reason

But I would really like to know and the entire YTuner user community would too. :wink: 
I suspect that perhaps the YTuner DNS service sets some properties incorrectly when intercepting requests and sending a response back to the AVR so that the NP-S2000 does not accept them, although most other AVRs work with it correctly... or simply YTuner DNS Service could not connect to the DNS server set in the `DNSServers` option of `ytuner.ini` file. :thinking:  Therefore, your tests of changing DNS servers and observing whether it causes YTuner to work correctly would be invaluable for us and future YTuner users.
What is/was your YTuner machine OS and CPU arch? Are you familiar with network sniffers like ... Wireshark?

>Do you plan to do on-the-fly transcoding, for example from AAC to MP3?

Well, as Beatrice mentioned, we already talked about this and this real-time transcoding process will require a lot of CPU processing power. I haven't abandoned this idea, especially since my RX-V671 does not support AAC and AAC+ while my other RX-V673 does. I can try it in my free time... see how it works. :thinking:  Yeah.... "free time"... it's getting harder to find something like that. :wink:

>ability to see the radio quality before or after the title, for example "[128] Dance Wave!" on my yamaha's display.

What purpose would this serve? If you want to filter out certain stations with unsupported codecs or bitrates use `avr.ini` settings. But .. as Beatrice mentioned before `Radiowrowser.info` database is not properly updated with correct codec/bitrate info.

> The only proper way is analyzing the stream data, but this is not implemented with ytuner and also not at vtuner or ycast.

Right. :+1: YTuner does not even participate in the process of downloading the audio stream itself by the AVR.

**Reply by Ice64zzz** [reply]
> What is/was your YTuner machine OS and CPU arch? 

At first I installed the YTuner on Windows x64 inside my local network, which was in NAT, specified it as DNS for, but the YTuner did not work. After that, I installed the YTuner on the VDS server on the Internet on Ubuntu 22.04 x64, and specified my router as DNS, in which I made a static host entry - everything worked.

> Are you familiar with network sniffers like ... Wireshark?

Yes, i am. But never used it in ubuntu console.

> real-time transcoding process will require a lot of CPU processing power
 
a lot is relative.. My old qnap TS-131 with ARM Cortex-A9 MPCore transcoding huge video-files with audio on-the-fly without any problem

> What purpose would this serve?
Some radio stations can play old songs in bad quality while having good overall quality. With [320] in the station name I could just wait for the next song. Some stations in [96] I would not want to filter because they have unique local news.

I noticed that some radiostations (50/50) are not playing on my np-s2000, for example MANGORADIO and Classic vinyl HD from most popular in Radio-browser. What can be reason of this? How can i find rule to hide/force to play them?

**Reply by coffeegreg** [reply]
> Yes, i am. But never used it in ubuntu console.

Great! Wireshark can help us. :rocket:  I would like to ask you to do a test with YTuner on your Windows x64 platform.
Scenario 1:
- Disable all your router DNS entries related to *ytuner.com and enable YTuner DNS Service with default 8.8.8.8 DNS Google service as `ytuner.ini` `DNSServers` setting entry.
- Set AVR's DNS Server to YTuner machine, restart AVR (power plug-off and plug-in)
- Start Wireshark and you will try to cach all DNS releated communication frames : AVR ---request---> YTuner DNS Service and AVR <---response---YTuner DNS Service
-  Start AVR and try to browse anything to back to this scenario with many log entries like this one :
> 04.11.2024 20:30:42 : Dbg : DNS Query intercept : radioyamaha.vtuner.com.

Scenario 2:
- Disable YTuner DNS Service, restart YTuner and restore all your current DNS settings (router,AVR) where everything works fine.
- Restart AVR (power plug-off and plug-in)
- Start/restart Wireshark again and you will try to cach the same all DNS releated communication frames : AVR ---request---> DNS Server and AVR <---response--- DNS Server
-  Start AVR and try to browse anything to capture DNS communication frames by Wireshark

After all let me know about all this captured DNS frames content in both scenarios (`cap`, `pcap`, `pcapng` files are most welcome). :rocket: 
I would like to see the differences between them to possibly improve/fix the YTuner DNS service. :+1: 

>  noticed that some radiostations (50/50) are not playing on my np-s2000, for example MANGORADIO 

Let's see ... Information about "MANGORADIO" from the `Radio-browser.info` :
http://all.api.radio-browser.info/json/stations/search?name=MANGORADIO
```
[{
.....
        "name": "MANGORADIO",
        "url": "http://stream.mangoradio.de/",
        "url_resolved": "https://mangoradio.stream.laut.fm/mangoradio?t302=2024-11-07_17-49-48&uuid=9d2e9103-2731-4ef6-a962-23b32ad67304",
        "homepage": "https://mangoradio.de/",
        "favicon": "",
        "tags": "mango,mangoradio,mongo,mongoradio,public radio,webradio",
        "country": "Germany",
        "countrycode": "DE",
        "iso_3166_2": null,
        "state": "",
        "language": "german",
        "languagecodes": "de",
        "votes": 711481,
        "lastchangetime": "2023-11-04 17:05:55",
        "lastchangetime_iso8601": "2023-11-04T17:05:55Z",
        "codec": "MP3",
        "bitrate": 128,
....
}]
```
Yup ... URL link seems valid, codec and bitrate too ... So, What the hell ... ? :imp: The answer is simple in this case ... 
Your NS-S200 does not support SSL/TLS (https). Don't worry. Most older AVRs don't support it, my two Yamahas don't either.
That's why there is a `Protocol` option in the `avr.ini` file with the possibility to set as :
`all-as-http = select all stations (http+https) and strip "s" from "https" URLs and use "http" instead.`
But, wait a minute .. this radio station URL is HTTP without secure SSL/TLS protocol : `http://stream.mangoradio.de` :exclamation: 
Yes, but ... let's see what's hidden deeper in this link. Let's use it :

![obraz](https://github.com/user-attachments/assets/17a200ec-2d9f-4d88-bfb9-1ca7c02d6090)

What do we have here? Redirect! ... even two! And where does it redirect?

![obraz](https://github.com/user-attachments/assets/4dfb0c3f-85ef-4142-88f0-572a885d947c)

..of course .. to the another HTTPS link :exclamation: :angry: 
Unfortunately, the entire operation takes place outside YTuner, only between the AVR and the website of the selected station. YTuner believes that everything is ok.
Now if you take this new redirected link `https://mongoradio.stream.laut.fm/...` and put it into the `station.ini` file YTuner will remove the "s" from the `https` link and send it to AVR and there is a big chance you will finally hear the music. At least for a while ... ;) Why? Because this link contains some “extra” parameters that are probably only valid for a while. Sometimes cookie support is required but our simple AVRs cannot handle it. :disappointed: One may ask: Why do they do this? The answer is just as simple ... If you don't know what it's about, then it's about money. :moneybag: Some radio stations operate thanks to money from advertisements that you see in your webbrowser window while listening to their music. They have no interest in you listening to their music on your AVR without watching their ads. I don't think it's like this everywhere, but it's common.

Finally, What can be done to deal with this? YTuner can use `url_resolved` link (most likely the best option but I'm not sure if it's always checked and filled) or I can add some background process thread to YTuner that would emulate a browser and check all links one by one and get the right link after redirection (Apparently the `radio-browser.info` does it). 
So... What to do now? Switch to another radio station and enjoy the music. :wink: :speaker: :notes: 

In addition to the above example, there are many others that cause our AVRs cannot play certain stations. Wrong codec, bitrate too high , etc. 
Phew... I wrote a lot and my English is not my strongest point but I hope everything is clear. :+1:

**Reply by coffeegreg** [reply]
Back to this "redirection" problem... I think I have an idea how to deal with it. :thinking: :bulb: 
Once I finish the current changes I will come back to this case. Hang on ... :rocket:

**Reply by Ice64zzz** [reply]
> Back to this "redirection" problem... I think I have an idea how to deal with it. 🤔 💡
> Once I finish the current changes I will come back to this case. Hang on ... 🚀

Many radio stations including Mangoradio started working on my Yamaha NP-S2000 after updating to YTuner v1.2.5. Thank you for solving this problem! 🚀

**Reply by coffeegreg** [reply]
Hi @Ice64zzz 
The problem is that I haven't done anything in this area yet. :eyes: I already have a concept and I will work on it as soon as possible but for now I haven't done or implemented anything. Lately I've been spending a lot of time helping people whose devices have recently stopped working because vTuner stopped supporting them and I didn't have time to deal with the new functionalities.
I want to solve the problem of link resolve and the fact that some devices don't support HTTP302 redirects. :rocket: 
I'm sorry, I do what I can but I also have some other work .. be patient. :+1:
